### PR TITLE
M10: add user authentication and authorisation.

### PR DESCRIPTION
**Pull request by Lucas-Fernandes-Nogueira**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -12,6 +12,8 @@ requests = "^2.24.0"
 gunicorn = "^20.0.4"
 pymongo = {extras = ["srv"], version = "^3.11.3"}
 mongomock = "^3.22.1"
+oauthlib = "^3.1.0"
```
Comment (BenRamchandani) [inline pyproject.toml]: I don't think you actually use this (which is fine, but you can remove it)

File: user.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    
+class User(UserMixin):
+    def __init__(self, id):
+        self.id = id
+
+    def has_reader_role(self):
+        return USER_ROLES[self.id] == 'reader'
+
+    def has_writer_role(self):
+        return USER_ROLES[self.id] == 'writer'
```
Comment (BenRamchandani) [inline user.py]: I suggest you do `USER_ROLES.get(self.id) == 'writer'` here, that way other people can use the app as readers. `dict['key']` throws if `'key'` isn't in `dict`, `dict.get('key')` returns `None`.

Also I don't think `has_reader_role` is used

**Review by BenRamchandani — CHANGES_REQUESTED** [review]
Hey Lucas,

Can you update your README + .env.template with the new setup steps?

Would also be good to make sure users that aren't in `user_roles.json` can use the app as readers. Not strictly required, but it makes it easier to mark 🙂

File: user.py (hunk 2 above)
Comment (Lucas-Fernandes-Nogueira) [reply user.py]: Thanks! Just changed it 😁.

**Comment by Lucas-Fernandes-Nogueira** [comment]
I've done all the markups now 😃.
Regarding "Would also be good to make sure users that aren't in user_roles.json can use the app as readers.", I think it's already doing that. I've tested deleting my GitHub id from user_roles.json and it let me access the app as a reader. Did you mean something else? Do I need to make that more explicit in the code?

**Comment by BenRamchandani** [comment]
> Regarding "Would also be good to make sure users that aren't in user_roles.json can use the app as readers.", I think it's already doing that. I've tested deleting my GitHub id from user_roles.json and it let me access the app as a reader. Did you mean something else? Do I need to make that more explicit in the code?

That's what the `USER_ROLES[self.id] == 'writer'` -> `return USER_ROLES.get(self.id) == 'writer'` change does. `dict["key"]` throws if the key doesn't exist (the user id is not in USER_ROLES), so the app wasn't useable before if you weren't in the list.

**Review by BenRamchandani — APPROVED** [review]
Looks good to merge.

Let me know once you've resubmitted it on Aptem and I'll accept it.

**Comment by Lucas-Fernandes-Nogueira** [comment]
> That's what the USER_ROLES[self.id] == 'writer' -> return USER_ROLES.get(self.id) == 'writer' change does. dict["key"] throws if the key doesn't exist (the user id is not in USER_ROLES), so the app wasn't useable before if you weren't in the list.
That's right, sorry for my confusion 😅.

**Pull request merged**